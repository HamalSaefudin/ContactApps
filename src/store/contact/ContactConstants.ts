export enum ContactActionType {
  SET_LAST_VISIT_CONTACT = 'SET_LAST_VISIT_CONTACT',
  UPDATE_LAST_VISIT_CONTACT = 'UPDATE_LAST_VISIT_CONTACT',
}
