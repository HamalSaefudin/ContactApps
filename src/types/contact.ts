export interface IContact {
  firstName: string;
  lastName: string;
  age: number;
  photo: string;
  id: string;
  callHistory?: string;
}

export interface ICreateContact {
  firstName: string;
  lastName: string;
  age: number;
  photo: string;
}
