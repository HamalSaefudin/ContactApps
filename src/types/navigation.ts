export enum routesEnum {
  HOME = 'HOME',
  DETAIL_CONTACT = 'DETAIL_CONTACT',
}

export type RoutesParamList = {
  [routesEnum.HOME]: undefined;
  [routesEnum.DETAIL_CONTACT]: {id: string};
};
