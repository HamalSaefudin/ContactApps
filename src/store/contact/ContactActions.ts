import {IContact} from '@/types/contact';
import {ContactActionType} from './ContactConstants';

export const setLastVisitContact = (payload: IContact) => ({
  type: ContactActionType.SET_LAST_VISIT_CONTACT,
  payload,
});

export const editLastVisitContact = (payload: IContact) => ({
  type: ContactActionType.UPDATE_LAST_VISIT_CONTACT,
  payload,
});
