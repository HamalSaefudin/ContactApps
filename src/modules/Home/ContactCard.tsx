import moment from 'moment';
import React from 'react';
import {Image, StyleSheet, Text, TouchableOpacity, View} from 'react-native';
import LinearGradient from 'react-native-linear-gradient';
import Animated, {FadeInDown} from 'react-native-reanimated';
import ShimmerPlaceholder from 'react-native-shimmer-placeholder';
import {moderateScale} from 'react-native-size-matters';

import Icon, {ICON_CATEGORY} from '@/components/Icons';
import Colors from '@/theme/Colors';
import Fonts from '@/theme/Fonts';
import Layout from '@/theme/Layout';
import {IContact} from '@/types/contact';
import {getImg} from '@/utils/utils';

const AnimatedTouchable = Animated.createAnimatedComponent(TouchableOpacity);

interface ICardProps {
  onPress: (item: IContact) => void;
  isLoading?: boolean;
  item: IContact;
  index: number;
}

const ContactCard: React.FC<ICardProps> = ({
  item,
  index,
  onPress,
  isLoading = true,
}) => {
  return (
    <AnimatedTouchable
      testID="contact-card"
      onPress={() => {
        onPress(item);
      }}
      entering={FadeInDown.delay(index * 100)}
      style={styles.cardContainer}>
      <View style={styles.cardContentContainer}>
        <ShimmerPlaceholder
          visible={!isLoading}
          LinearGradient={LinearGradient}
          shimmerStyle={styles.cardImgContainer}>
          <View style={styles.cardImgContainer}>
            <Image
              testID="img-contact-card"
              source={{uri: getImg(item.photo)}}
              style={styles.cardImg}
            />
          </View>
        </ShimmerPlaceholder>
        <View style={styles.textContainer}>
          <View>
            <ShimmerPlaceholder
              visible={!isLoading}
              LinearGradient={LinearGradient}>
              <Text testID="contact-card-text-name" style={styles.txtName}>
                {`${item.firstName} ${item.lastName}`}
              </Text>
            </ShimmerPlaceholder>
            <ShimmerPlaceholder
              visible={!isLoading}
              LinearGradient={LinearGradient}>
              <Text testID="contact-card-text-age" style={styles.txtAge}>
                {`${item.age} Years Old`}
              </Text>
            </ShimmerPlaceholder>
          </View>
          <ShimmerPlaceholder
            visible={!isLoading}
            LinearGradient={LinearGradient}>
            <Text
              testID="contact-card-call-history"
              style={styles.txtCallHistory}>
              {moment(item.callHistory).format('MMMM DD, YYYY')}
            </Text>
          </ShimmerPlaceholder>
        </View>
      </View>
      <ShimmerPlaceholder
        visible={!isLoading}
        LinearGradient={LinearGradient}
        shimmerStyle={styles.iconArrowShimmer}>
        <Icon category={ICON_CATEGORY.ANT_DESIGN} name="rightcircle" />
      </ShimmerPlaceholder>
    </AnimatedTouchable>
  );
};

export default ContactCard;

const styles = StyleSheet.create({
  cardContainer: {
    padding: 16,
    borderBottomWidth: 1,
    borderColor: Colors.common.border,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  cardContentContainer: {
    flexDirection: 'row',
    gap: moderateScale(15),
  },
  textContainer: {
    justifyContent: 'space-between',
  },
  iconArrowShimmer: {
    width: moderateScale(16),
    height: moderateScale(16),
    borderRadius: moderateScale(8),
  },
  cardImgContainer: {
    width: moderateScale(52),
    height: moderateScale(52),
    borderRadius: moderateScale(26),
    backgroundColor: Colors.common.border,
  },
  cardImg: {
    ...Layout.fullSize,
    borderRadius: moderateScale(26),
  },
  txtName: {
    ...Fonts.bodyBold.body1,
    color: Colors.typography.primary,
  },
  txtAge: {
    ...Fonts.bodyBold.body3,
    color: Colors.typography.secondary,
  },
  txtCallHistory: {
    ...Fonts.bodyRegular.body3,
    color: Colors.typography.subtitle,
  },
});
