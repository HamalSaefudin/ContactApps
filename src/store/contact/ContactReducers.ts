import {IContact} from '@/types/contact';
import {ContactActionType} from './ContactConstants';

interface ContactState {
  lastVisitContact: IContact[];
}

const contactInitialState: ContactState = {
  lastVisitContact: [],
};

interface ContactAction {
  type: keyof typeof ContactActionType;
  payload: IContact;
}

export const ContactReducers = (
  state: ContactState = contactInitialState,
  action: ContactAction,
): ContactState => {
  const {payload, type} = action;

  switch (type) {
    case ContactActionType.SET_LAST_VISIT_CONTACT:
      const lastContact =
        state.lastVisitContact[state.lastVisitContact.length - 1];
      if (lastContact && lastContact.id === payload.id) {
        return state;
      }
      return {
        ...state,
        lastVisitContact: [...state.lastVisitContact.splice(-4), payload],
      };
    case ContactActionType.UPDATE_LAST_VISIT_CONTACT:
      const idx = state.lastVisitContact.findIndex(
        contact => contact.id === payload.id,
      );

      if (idx >= 0) {
        const updatedContacts = state.lastVisitContact.map((contact, index) =>
          index === idx ? payload : contact,
        );
        return {
          ...state,
          lastVisitContact: updatedContacts,
        };
      }
      return state;
    default:
      return state;
  }
};
