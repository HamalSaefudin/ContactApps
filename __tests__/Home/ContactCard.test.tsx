import ContactCard from '@/modules/Home/ContactCard';
import {IContact} from '@/types/contact';
import {fireEvent, render} from '@testing-library/react-native';
import moment from 'moment';
import React from 'react';
import {act} from 'react-test-renderer';

const mockContact: IContact = {
  id: '1',
  firstName: 'John',
  lastName: 'Doe',
  age: 30,
  photo: 'https://example.com/john.jpg',
  callHistory: moment().subtract(1, 'days').toISOString(),
};

describe('ContactCard', () => {
  it('should match snapshot', () => {
    const component = render(
      <ContactCard
        item={mockContact}
        index={0}
        onPress={() => {}}
        isLoading={false}
      />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
  it('renders correctly when not loading', () => {
    const component = render(
      <ContactCard
        item={mockContact}
        index={0}
        onPress={() => {}}
        isLoading={false}
      />,
    );

    const textName = component.getByTestId('contact-card-text-name');
    const textAge = component.getByTestId('contact-card-text-age');
    const image = component.getByTestId('img-contact-card');
    const textCallHistory = component.getByTestId('contact-card-call-history');

    expect(image).toBeTruthy();
    expect(textName).toBeTruthy();
    expect(textName.children[0]).toEqual(
      mockContact.firstName + ' ' + mockContact.lastName,
    );
    expect(textAge).toBeTruthy();
    expect(textAge.children[0]).toEqual(mockContact.age + ' Years Old');
    expect(textCallHistory).toBeTruthy();
    expect(textCallHistory.children[0]).toEqual(
      moment(mockContact.callHistory).format('MMMM DD, YYYY'),
    );
  });

  it('onPressCalled when the card is pressed', () => {
    const handlePress = jest.fn();
    const component = render(
      <ContactCard
        item={mockContact}
        index={0}
        onPress={handlePress}
        isLoading={false}
      />,
    );

    const card = component.getByTestId('contact-card');
    act(() => {
      fireEvent.press(card);
    });
    expect(handlePress).toHaveBeenCalledWith(mockContact);
  });
});
