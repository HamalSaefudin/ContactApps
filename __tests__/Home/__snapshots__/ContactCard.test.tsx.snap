// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactCard should match snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "borderBottomWidth": 1,
      "borderColor": "#D9D9D9",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "opacity": 1,
      "padding": 16,
    }
  }
  testID="contact-card"
>
  <View
    style={
      {
        "flexDirection": "row",
        "gap": 23.57142857142857,
      }
    }
  >
    <View
      style={
        [
          false,
          {
            "overflow": "hidden",
          },
          false,
          undefined,
        ]
      }
    >
      <View
        style={
          [
            false,
            undefined,
          ]
        }
      >
        <View
          style={
            {
              "backgroundColor": "#D9D9D9",
              "borderRadius": 40.85714285714286,
              "height": 81.71428571428572,
              "width": 81.71428571428572,
            }
          }
        >
          <Image
            source={
              {
                "uri": "https://example.com/john.jpg",
              }
            }
            style={
              {
                "borderRadius": 40.85714285714286,
                "height": "100%",
                "width": "100%",
              }
            }
            testID="img-contact-card"
          />
        </View>
      </View>
    </View>
    <View
      style={
        {
          "justifyContent": "space-between",
        }
      }
    >
      <View>
        <View
          style={
            [
              false,
              {
                "overflow": "hidden",
              },
              false,
              undefined,
            ]
          }
        >
          <View
            style={
              [
                false,
                undefined,
              ]
            }
          >
            <Text
              style={
                {
                  "color": "#26245D",
                  "flexShrink": 1,
                  "fontFamily": "Roboto-Bold",
                  "fontSize": 22,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                }
              }
              testID="contact-card-text-name"
            >
              John Doe
            </Text>
          </View>
        </View>
        <View
          style={
            [
              false,
              {
                "overflow": "hidden",
              },
              false,
              undefined,
            ]
          }
        >
          <View
            style={
              [
                false,
                undefined,
              ]
            }
          >
            <Text
              style={
                {
                  "color": "#5C81F5",
                  "flexShrink": 1,
                  "fontFamily": "Roboto-Bold",
                  "fontSize": 15.714285714285714,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                }
              }
              testID="contact-card-text-age"
            >
              30 Years Old
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          [
            false,
            {
              "overflow": "hidden",
            },
            false,
            undefined,
          ]
        }
      >
        <View
          style={
            [
              false,
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "#898898",
                "flexShrink": 1,
                "fontFamily": "Roboto-Regular",
                "fontSize": 15.714285714285714,
                "fontStyle": "normal",
                "fontWeight": "normal",
              }
            }
            testID="contact-card-call-history"
          >
            June 08, 2024
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      [
        false,
        {
          "overflow": "hidden",
        },
        false,
        undefined,
      ]
    }
  >
    <View
      style={
        [
          false,
          undefined,
        ]
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#000000",
              "fontSize": 25.142857142857142,
            },
            undefined,
            {
              "fontFamily": "anticon",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        î˜ˆ
      </Text>
    </View>
  </View>
</View>
`;
